### PR TITLE
Speed up batched note decryption

### PR DESCRIPTION
**Pull request by str4d**

### CODE REVIEW

**Review by daira — APPROVED** [review]
utACK

File: src/keys.rs [hunk 1]
```diff
@@ -567,15 +567,34 @@ impl SharedSecret {
         self.0.to_bytes()
     }
 
+    /// Only for use in batched note encryption.
+    pub(crate) fn batch_to_affine(shared_secrets: &[Option<Self>]) -> Vec<pallas::Affine> {
+        let secrets: Vec<_> = shared_secrets
+            .iter()
+            .filter_map(|s| s.as_ref().map(|s| *(s.0)))
+            .collect();
+        let mut secrets_affine = vec![pallas::Affine::identity(); shared_secrets.len()];
```
Comment (str4d) [inline src/keys.rs:576]: This should be `secrets.len()` (as the two slices passed into `batch_normalize` need to be the same length).

File: src/keys.rs [hunk 2]
```diff
@@ -567,15 +567,34 @@ impl SharedSecret {
         self.0.to_bytes()
     }
 
+    /// Only for use in batched note encryption.
+    pub(crate) fn batch_to_affine(shared_secrets: &[Option<Self>]) -> Vec<pallas::Affine> {
```
Comment (str4d) [inline src/keys.rs:571]: I'll see if I can make this return an `impl Iterator`, so I can move the interpolation logic into here for clarity.

**Comment by daira** [comment]
Typo in commit comment:
> This makes the method interface clearer, as the same patter of shared
> secrets is returned as was provided.

s/patter/pattern/

**Review by daira — APPROVED** [review]
utACK

**Pull request merged**